addon/model.js

```diff
@@ -643,12 +643,21 @@ export default class MegamorphicModel extends EmberObject {
     schemaInterface._suppressNotifications = priorSuppressNotifications;
 
     const hasDirtyAttr = recordData.hasChangedAttributes();
-    const isDirty = get(this, 'isDirty');
+    if (!CUSTOM_MODEL_CLASS) {
+      const isDirty = get(this, 'isDirty');
 
-    if (hasDirtyAttr && !isDirty) {
-      this._updateCurrentState(!CUSTOM_MODEL_CLASS && updatedUncommitted);
-    } else if (!hasDirtyAttr && isDirty) {
-      this._updateCurrentState(!CUSTOM_MODEL_CLASS && loadedSaved);
+      if (hasDirtyAttr && !isDirty) {
+        this._updateCurrentState(updatedUncommitted);
+      } else if (!hasDirtyAttr && isDirty) {
+        this._updateCurrentState(loadedSaved);
+      }
+    } else {
+      const cachedIsDirty = this._topModel._isDirty;
+
+      // `isDirty` cached value does not match our current one -> we need to update our state
+      if (hasDirtyAttr !== cachedIsDirty) {
+        this._updateCurrentState();
+      }
```

Comment:
Can we swap the order here (i.e. change the conditional to if `CUSTOM_MODEL_CLASS`) so it matches the order where we set the cached value.

Having the then/else branches switched makes this harder to read.

Reply:
Done